### PR TITLE
Add CIFS and Panasas support to storage promise type

### PR DESCRIPTION
**Pull request by nickanderson**
The PR this code came from: https://github.com/cfengine/core/pull/3362
Shared by https://github.com/favoritelotus

O-pened in new PR because PR was not clean.

Changelog: TODO

### CODE REVIEW

**Comment by cf-bottom** [comment]
Thank you for submitting a PR! Maybe @vpodzime can review this?

**Comment by favoritelotus** [comment]
> @favoritelotus It would be great if you could share any other information about your usage with this as well.

Hi Nick,

Sure. I have prepared example policy files with the comments for both cifs and panfs. You will find them in the attached zip archive.
[examples.zip](https://github.com/cfengine/core/files/2498177/examples.zip)

Regards,
Pavel.

File: cf-agent/nfs.c [hunk 1]
```diff
@@ -126,13 +127,36 @@ static const char *const VMOUNTOPTS[] =
     [PLATFORM_CONTEXT_ANDROID] = "defaults",         /* android */
 };
 
+
+void GetHostAndSource(char *buf, char *host, char *source)
```
Comment (vpodzime) [inline cf-agent/nfs.c]: `buf` should be `const char *`.

File: cf-agent/nfs.c [hunk 2]
```diff
@@ -126,13 +127,36 @@ static const char *const VMOUNTOPTS[] =
...
+/* Extracts from buffer host & source of cifs and panfs network systems */
+{
+    int i = 0, i2 = 0, bs = 0;
+
+    // copy host
+    while(bs != 3)
```
Comment (vpodzime) [inline cf-agent/nfs.c]: Would be great to add an example of the data this should process.

File: cf-agent/nfs.c [hunk 3]
```diff
@@ -223,7 +255,7 @@ bool LoadMountInfo(Seq *list)
 
 
 #if defined(__sun) || defined(__hpux)
-        if (IsAbsoluteFileName(buf3))
+        if (IsAbsoluteFileName(buf3) && !cifs)
```
Comment (vpodzime) [inline cf-agent/nfs.c]: Why are these checks needed?

File: cf-agent/nfs.c [hunk 4]
```diff
@@ -126,13 +127,36 @@ static const char *const VMOUNTOPTS[] =
     [PLATFORM_CONTEXT_ANDROID] = "defaults",         /* android */
 };
 
+
+void GetHostAndSource(char *buf, char *host, char *source)
+/* Extracts from buffer host & source of cifs and panfs network systems */
+{
+    int i = 0, i2 = 0, bs = 0;
```
Comment (olehermanse) [inline cf-agent/nfs.c]: So `bs` is used to count slashes, and `i2` is used as an index for `source`. I would call these variables `slashes` and `source_index`, respectively.

File: cf-agent/nfs.c [hunk 5]
```diff
@@ -567,7 +615,19 @@ PromiseResult VerifyMount(EvalContext *ctx, char *name, const Attributes *a, con
     PromiseResult result = PROMISE_RESULT_NOOP;
     if (!DONTDO)
     {
-        snprintf(comm, CF_BUFSIZE, "%s -o %s %s:%s %s", CommandArg0(VMOUNTCOMM[VSYSTEMHARDCLASS]), opts, host, rmountpt, mountpt);
+
+        if (strcmp(a.mount.mount_type, "panfs") == 0)
```
Comment (olehermanse) [inline cf-agent/nfs.c]: `a` has been changed to a pointer, so this needs to be updated to `a->mount`. (Maybe @nickanderson 's mistake while cherry-picking.

**Comment by favoritelotus** [comment]
Hello Vratisalv,

The variable bs stands for back slash. Inside the loop, the variable simply
counts slashes to find where the host name ends and source begins when a
CIFS share is mounted as //server/source. The same applies for panfs, which
has a similar mount syntax, e.g. panfs://server/source.

Since CIFS mounts start with two slashes, without the check "if
(IsAbsoluteFileName(buf3) && !cifs)" they will be interpreted as local
mounts and transformed to  localhost://server/host.

Regards,
Pavel.

File: cf-agent/nfs.c [hunk 6]
```diff
@@ -47,6 +47,7 @@
 static int FSTAB_EDITS = 0; /* GLOBAL_X */
 static Item *FSTABLIST = NULL; /* GLOBAL_X */
 
+void GetHostAndSource(char *buf, char *host, char *source);
```
[the remark this answers is not in the document]
Comment (nickanderson) [reply cf-agent/nfs.c]: @favoritelotus  Do you plan to come back around and make fix-ups per these suggestions?

**Comment by vpodzime** [comment]
Nice! Thanks for the detailed explanation! Could you please address the code issues pointed out in the reviews?

**Comment by olehermanse** [comment]
@nickanderson @vpodzime It looks like @favoritelotus is not responding. We should either fix this one ourself, or close it.

**Comment by favoritelotus** [comment]
Hello,
Sorry for not replying long. I don't have much free time to 
I just want to share my  new patch, which fixes  some bugs related to sigfault caused by Panasas mounts. The patch was debugged, tested and finally put into production. 
We have been using successfully CFengine 3.12.0 with my patch in production for three mounts on 194 servers. The patch is attached with 

[panfs-cifs-patch.zip](https://github.com/cfengine/core/files/3694422/panfs-cifs-patch.zip)

sample policy file.

**Comment by olehermanse** [comment]
@favoritelotus Thank you for your reply, I'll take over this one. See this PR for further progress: https://github.com/cfengine/core/pull/3893

**Comment by favoritelotus** [comment]
Hello Ole,

Thank you. I will release another patch soon, which adds support for SSHFS
to the storage promise type.

Regards,
Pavel.

**Pull request closed without merging**